### DISCUSSION TITLE
cmdline when booting EFI image type

### DISCUSSION THREAD

**Discussion by dukovac** [ipxe/ipxe#1367] [Q&A] [answered]
Hello!

I am curious as to why the command line for an efi image is a concatenation of the image name and image command line, rather than just the image command line? https://github.com/ipxe/ipxe/blob/master/src/image/efi_image.c#L203

The code will set the image's LoadOptions data to this newly concatenated cmdline (name+cmdline) before passing control to the newly loaded image. https://github.com/ipxe/ipxe/blob/master/src/image/efi_image.c#L263

When booting a UKI built using `linuxx64.efi.stub` and the ipxe conf `#!ipxe; kernel http://path/to/uki.efi; boot` I run into the issue that LoadOptions contains the `uki.efi` filename. This non empty LoadOptions is then passed to the kernel rather than the cmdline baked into `uki.efi`. If LoadOptions is null, the baked in cmdline is extracted and used.

**Comment by mcb30** [comment, marked as answer]
There are no well-defined standards for UEFI command lines. (It's not even defined to be a string: it's an opaque binary blob and some loaders will pass bizarre binary structures).

Passing the image name and command line is just what iPXE has always done, and was almost certainly based on the command line as constructed by some quasi-definitive reference such as the UEFI shell at the time. 

It could not be changed now without breaking backwards compatibility, so this is a fix that would have to go into the UKI loader rather than into iPXE, sorry.

**Reply by NiKiZe** [reply]
And can we also assume that this is how efi shell and some other implementations does it as well? Would it be worth having a separate page documenting which does what?

**Reply by mcb30** [reply]
Some history and background captured here: https://github.com/systemd/systemd/pull/35608#issuecomment-2554126609
